Hw4 coroutines load data

evkingen opened the pull request:
![lXN3NOcmbls](https://user-images.githubusercontent.com/38916658/104127536-a89ae500-5373-11eb-88dc-da4071fae19b.jpeg)
![qnY5YcZZbKo](https://user-images.githubusercontent.com/38916658/104127537-ab95d580-5373-11eb-9c6f-149558ae026e.jpeg)
![-3VyObu4DtY](https://user-images.githubusercontent.com/38916658/104127423-0bd84780-5373-11eb-92a8-6cb999dc7c5a.jpeg)
![ScHDV8BHKCc](https://user-images.githubusercontent.com/38916658/104127425-0f6bce80-5373-11eb-8999-829cc70b8372.jpeg)
![CSaN4BoomOU](https://user-images.githubusercontent.com/38916658/104127432-1266bf00-5373-11eb-88e3-87987f9d690a.jpeg)
![S7303266](https://user-images.githubusercontent.com/38916658/104127437-1561af80-5373-11eb-9ab4-a74d231f5a76.JPG)
![_-5h1dYfJ00](https://user-images.githubusercontent.com/38916658/104127439-17c40980-5373-11eb-9264-1d535a007ac6.jpeg)
![y4x47LMpp84](https://user-images.githubusercontent.com/38916658/104127443-1a266380-5373-11eb-9969-c68f60cddeb1.jpeg)
![cnddilgLk08](https://user-images.githubusercontent.com/38916658/104127446-1c88bd80-5373-11eb-944b-2f63d88685b6.jpeg)

evkingen committed:
Добавлены зависимости kotlin serialization plugin и корутины
В assets Добавлены данные о фильмах и актерах в формате json а также парсер данных.
Изменена модель.

evkingen committed:
Refactoring XML и правки по дизайну

evkingen committed:
MainActivity больше не реализует интерфейс FragmentMoviesList.OnMovieCardClickListener
Adapter - рефакторинг
Во фрагменты добавлена загрузка данных из assets (метод startLoadMovies) + рефакторинг

DmitryKolpo commented on app/src/main/AndroidManifest.xml:
```diff
@@ -1,14 +1,17 @@
 
+    <uses-permission android:name="android.permission.INTERNET" />
     <application
         android:allowBackup="true"
```
обычно убирают разрешение на бэкап. Человек из редмедроботов про безопасность рассказывал, что сильно проще реверсерам становится, если бэкап включен

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/MainActivity.kt:
```diff
@@ -6,26 +6,31 @@ import com.alohagoha.aaaa.databinding.ActivityMainBinding
+                .commit()
         }
         setContentView(mainBinding.root)
+        (supportFragmentManager.findFragmentByTag(FragmentMoviesList.FRAGMENT_TAG) as FragmentMoviesList?)?.clickListener =
+            onMovieCardClickListener
```
Эти строчки вызывают подозрения.
Что именно неправильно и как именно правильно сделать, сейчас не скажу.

Когда нам приходится искать что-то и явно кастить, то это означается, что есть некоторые предположения, которые потом могут оказаться не правдой и будет баг/ошибка.  Например, тут предполагается, что если стейт уже есть, то скорее всего там будет мупилист фрагмент. Как будто такие предположения хочется проверять во время компиляции, а не в рантайме.

Мини нюанс: можно использовать "as? Class" вместо "as Class?"

evkingen replied:
Переписал, решил поиграться с использованием FactoryFragment. Какую то постоянную штуку я понимаю как передать. А вот MovieId которые постоянно может быть разные в фабрике не понимаю как использовать.

А зачем Артур вообще на воркшопах рассказывал про этот метод? какой кейс у findFragmentByTag и findFragmentById, что они принимают и возвращают знаю, а в каком случае с ними можно работать? В навигации может пригодиться?

DmitryKolpo replied:
Когда чуть более общие/абстрактные задачи необходимо решать(да, например, свою кастомную навигацию писать), то там не обойтись без этих методов.

Суть в том, что тут не хэндлятся ситуации, когда фрагмент не найден или найден, но другого типа.
Но в нашем случае это и не надо, потому что мы знаем, что будет фрагмент списка фильмов. И в итоге можем перевести эту проверку из рантайма в компиляцию.
...
Фрагмент фабрика выглядит хорошо, код сильно лучше становится.

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/data/JsonLoad.kt:
```diff
@@ -0,0 +1,111 @@
+}
+
+@Suppress("unused")
+internal suspend fun loadMovies(context: Context): List<Movie> = withContext(Dispatchers.IO) {
+    val genresMap = loadGenres(context)
+    val actorsMap = loadActors(context)
+
+    val data = readAssetFileToString(context, "data.json")
+    parseMovies(data, genresMap, actorsMap)
+}
```
Давай данный(дата классы) и методы работы с ними раскидаем по разным файликам/классам.
Полноценно раскидывать по идее будешь в следующей домашке про архитектуру, но тут все равно хочется разделить, чтобы удобнее было читать и понимать где что.

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/data/JsonLoad.kt:
```diff
@@ -0,0 +1,111 @@
+                numberOfRatings = jsonMovie.votesCount,
+                minimumAge = if (jsonMovie.adult) 16 else 13,
+                runtime = jsonMovie.runtime,
+                genres = jsonMovie.genreIds.map {
+                    genresMap[it] ?: throw IllegalArgumentException("Genre not found")
+                },
+                actors = jsonMovie.actors.map {
+                    actorsMap[it] ?: throw IllegalArgumentException("Actor not found")
+                }
```
В целом, как я понимаю, бывают два способа:
один - если что-то идет не так, то крашить как можно раньше
другой - стараться, чтобы апа продолжила работать несмотря ни на что

Выбор зависит от приоритетов: либо нужна надежность в консистентности данных, либо нужна безотказная работа.

Тут получается, что если жанры или актеры не спарсились, то весь метод получения списка зафейлится.
Если сразу ломать метод - это твое осознанное решение, то ок.
Я бы тут лучше забивал на актеров и жанры и выводил то, что смогло спарсится.

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/FragmentMoviesDetails.kt:
```diff
@@ -1,35 +1,37 @@
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
+        _moviesDetailsBinding = FragmentMoviesDetailsBinding.bind(view)
+        movieId =
+            (savedInstanceState?.getInt(MOVIE_ID, NO_ID) ?: arguments?.getInt(MOVIE_ID, NO_ID))!!
```
Стоит стараться избегать "!!". Компилятор не запрещает не отправлять аргументы в этот фрагмент и стоит обработать кейс, когда аргументов нет. Хотя бы завершить приложение с каким-нибудь тоастом с информацией что происходит, а не просто краш.

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/FragmentMoviesDetails.kt:
```diff
@@ -49,8 +51,53 @@ class FragmentMoviesDetails : Fragment() {
+
+            if (movie.actors.isNotEmpty()) {
+                castTitleTv.visibility = View.VISIBLE
+                (castRv.adapter as ActorsListAdapter).updateList(movie.actors)
```
Лучше адаптер вынести в переменную. Тогда не придется тут кастовать.

Кастовать, проверять на инстанс класса - это всегда code smell. Если нет явных причин так делать, то скорее всего что-то пошло не так и надо переписать, чтобы не пришлось кастовать.

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/FragmentMoviesDetails.kt:
```diff
@@ -49,8 +51,53 @@ class FragmentMoviesDetails : Fragment() {
+
+
+            } else {
+                castTitleTv.visibility = View.INVISIBLE
```
В теории есть isVisible расширение в котлине, можно его использовать. Местами может быть удобнее.

castTitleTv.isVisible = movie.actors.isNotEmpty()

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/FragmentMoviesDetails.kt:
```diff
@@ -49,8 +51,53 @@ class FragmentMoviesDetails : Fragment() {
+    fun startLoadMovies() {
+        lifecycleScope.launchWhenCreated {
+            moviesDetailsBinding.progressBar.visibility = View.VISIBLE
+            loadMovies(context!!).find { it.id == movieId }?.let {
```
Везде стоит избавиться от "!!"

DmitryKolpo commented on app/src/main/res/layout/view_holder_movie.xml:
```diff
@@ -51,55 +51,54 @@
         <androidx.appcompat.widget.AppCompatTextView
             android:id="@+id/genre_tv"
-            android:layout_width="wrap_content"
+            android:layout_width="0dp"
             android:layout_height="wrap_content"
             android:layout_marginBottom="@dimen/padding_tiny"
-            android:maxLines="1"
-            android:text="@string/genre"
+            android:gravity="start"
+            android:maxLines="2"
             android:textAppearance="@style/TextAppearanceMaterialCustom.Body"
             android:textColor="@color/radical_red"
             android:textSize="@dimen/text_size_body_tiny"
             app:layout_constraintBottom_toTopOf="@id/count_review_tv"
+            app:layout_constraintEnd_toEndOf="@id/end_margin_guideline"
             app:layout_constraintStart_toStartOf="@id/start_margin_guideline" />
```
В лейауте удобно использовать tools атрибут, чтобы было примерно понятно где какой текст будет, сколько примерно занимать.

tools:text="Какой-то там текст"

evkingen replied:
на все пустые textView установил этот атрибут. Значения поискал в @tools:sample и не найдя ничего подходящего завел свои ресурсы.
На RecyclerView для актеров listItem пробовал устанавливать, но я так понял он не учитывается layoutManager что логично. Есть ли смысл устанавливать если всё равно искаженное представление имеешь) наверное нет)

DmitryKolpo replied:
Про свои ресурсы: можно так, но я не вижу ничего плохого, если прямо в лейауте хардкодом будет написано в tools что-то. Например, tools:text="Some looooong text".

Про rv не понял. Лейаут менеджер можно указать в лейауте через атрибут layoutManager. Точно также через tools его можно прописать. Не сработает только если у тебя сильно кастомный лейаут менеджер.

Смысл от этих атрибутов только в том, чтобы помогать тебе понимать, что тут примерно будет. Если какие-то тулзовые атрибуты нее помогают или наоборот дают неверное представление, то лучше не писать их.

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/FragmentMoviesDetails.kt:
```diff
@@ -49,8 +51,53 @@ class FragmentMoviesDetails : Fragment() {
+            storylineDescrTv.text = movie.overview
+            movieRating.rating = (movie.ratings * 5 / 10)
+            movieRestrictionRating.restrictionRating.text =
+                context!!.resources.getString(R.string.movie_restriction, movie.minimumAge)
```
можно сразу обращаться к resources (внутри фрагмента есть метод getResources)

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/FragmentMoviesList.kt:
```diff
@@ -66,12 +61,11 @@ class FragmentMoviesList : Fragment() {
+    companion object {
+        const val FRAGMENT_TAG = "TAG:FragmentMoviesList"
     }
 }
```
публичный/приватные методы/переменные нужно расположить в одном определенном порядке. Обычно чем публичнее, тем выше, чтобы сразу видеть внешнюю "кухню" класса, и не видеть внутреннюю

evkingen replied:
Так ну здравый смысл мне подсказывает что companion object и interface довольно публичные. Так лучше с них начинать? Странно а я в примерах в воркшопах точно где-то видел что companion object внизу располагали) 

А еще вот наткнулся...
https://kotlinlang.org/docs/reference/coding-conventions.html#source-code-organization
 Do not sort the method declarations alphabetically or by visibility, and do not separate regular methods from extension methods. Instead, put related stuff together, so that someone reading the class from top to bottom can follow the logic of what's happening.

Но мне не хватает опыта, потому что какое-то очень общая такая рекомендация...

DmitryKolpo replied:
Сорян, написал слишком поверхностно, предполагая, что "все всё знают".

Я не имел в виду только по публичности все отсортировать.
По ссылке как раз пишут про рекомендации порядка в классе: конструктор, переменные, методы, компаньон.
А уже каждый из блоков сортируют по видимости обычно, т.е.
public val/var
private val/var
init
public fun
private fun
companion
class

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/FragmentMoviesList.kt:
```diff
@@ -1,50 +1,44 @@
+class FragmentMoviesList : Fragment(R.layout.fragment_movies_list) {
 
-    private var clickListener: OnMovieCardClickListener? = null
+    var clickListener: OnMovieCardClickListener? = null
```
А нельзя тут сразу определить листенер, используя fragmentManager, чтобы не устанавливать листенер из активити?

К сожалению, нет времени самому проверить, но вроде это должно сработать.

evkingen replied:
Да, точно. Можно было бы так сделать через parentFragmentManager во фрагменте. Но разве это хорошая практика? Если бы каждый фрагмент управлял своими переходами?

У меня было  с самого начало сделано классическое общение через интерфейс и onAttach/onDettach во фрагменте я его прекрасно понимаю. Но получается, что в теории активити может начать реализовывать несколько интерфейсов. Это не противоречит Single Responsobility или реализация интерфейсов здесь в данном случае это "роутер" которым выступает активити?

Но я так понимаю язык нам позволяет не передавать активити во фрагмент, а передать лямбду?

Реализация типа 
if(saveInstanceState!=null){
    supportFragmentManaget.getFindFragmentByTag(FragmentMoviesList.FRAGMENT_TAG) as? FragmentMoviesList
}
я видел у нас на лекции по Фрагментам, Артур рассказывал что ситуации бывают разные и можно встретить и такой код. 
И мне показалось что в чатиках когда делали домашку на фрагменты обсуждали два вариант
1) общение через Interface 
2) Вот этот стремный способ. В итоге я больше в коде(в воркшопах) такой код нигде ни у кого не видел

Как вообще в простом виде можно сделать близкое к правильному? Всё таки использовать activity в качестве роутера с определенными showAllMovies() showDetailsMovie() ? А методы вынести в интерфейс и в Activity реализовать этот интерфейс?

DmitryKolpo replied:
Возможно я не совсем точно выразился.

Меня смущало, что листенер определен как public var. Когда я вижу публичную переменную, то ожидаю, что она нужна кому-то снаружи и что она меняется несколько раз. А на деле это поле один раз инициализируется и больше никогда не меняется, и нужно только для внутренних нужд.

Сейчас этот листенер передается в конструторе и это мне нравится.

Про навигацию, роутеры и SRP не буду тут рассказывать(да и не уверен, что смогу), тк вопрос у меня был в другом. Но если тебе сильно интересно, то в личку напиши, можем обсудить.

DmitryKolpo commented on app/src/main/java/com/alohagoha/aaaa/ui/rv/adapters/MoviesListAdapter.kt:
```diff
@@ -48,39 +88,24 @@ class MoviesListAdapter(private val context: Context,
-                movieRestrictionRating.restrictionRating.text = movie.restriction_rating
+                genreTv.text = movie.genres.joinToString { it.name }
+                movieNameTv.text = movie.title
+                movieRating.rating = (movie.ratings * 5 / 10)
```
В нескольких местах используется магические числа 5 и 10. Я бы сделал где-нибудь метод, который возвращает в нужном формате рейтинг.

evkingen replied:
Я тогда уберу это преобразование в Mapper который мапит полученную из JsonMovie в нашу сущность Movie в JsonLoader'e. Я уяснил что правила преобразования  между слоями таковы, что внутренний уровень абстракции получает данные в наилучше подходящем для него виде(чтобы ему не мапить их).  

И вроде бы у нас нет никаких взаимодействий с рейтингом кроме его отображение, так что подойдет такой формат, который в сущности бизнес-логики.

А вообще я тебя вот спрашивал как раз похожее пару PR назад
Вот смотри например  
      movieRestrictionRating.restrictionRating.text =  getString(R.string.movie_restriction, movie.minimumAge)
тоже пишется установка два раза и хотелось её куда-то вынести в какую-то умную вещь...

DmitryKolpo replied:
Про слои: да, все верно, хочется получать уже максимально готовые данные, чтобы брать и использовать их(тут могут быть нюансы с ресурсами на обсуждение, но общее правило верное).

Про рейтинг: да, тут достаточно понятное и простое место для преобразования, мне нравится.

Про возраст: это как раз нюанс с ресурсами. Для меня это "вкусовщина": считаю, что и текущий подход норм, и подход, где в адаптер приходит готовый String тоже ок. В каждом варианте есть плюсы и минусы свои.
По умолчанию я бы писал также, как у тебя - несколько getString в каждом адаптере.

DmitryKolpo reviewed (changes requested):
Нужно методы/переменные приватные/публичные в одном порядке разместить, чтобы не было в перемешку все.

Нужно избавиться от всех "!!".

evkingen committed:
Исправление замечаний по PR #4
-добавления tools:text к макетам.

evkingen committed:
Исправление по замечаниям PR #4
-Разбиение монолитного JsonLoad на классы dto и сущность JsonLoader по парсингу данных.

evkingen committed:
Исправление замечаний по PR #4
Добавлена фабрика MoviesFragmentFactory для фрагмента FragmentMoviesList
Создание и восстановление состояния фрагмента осуществляется через фабрику.
+рефакторинг кода.

DmitryKolpo reviewed (approved):
Стоит поменять очередность полей/методов внутри класса в соответствии с конвенциями(сорян, что запутал). А все остальное хорошо стало.
